pkg/controller/operators/catalog/operator.go

```diff
@@ -380,10 +380,22 @@ func NewOperator(ctx context.Context, kubeconfigPath string, clock utilclock.Clo
 	op.lister.RbacV1().RegisterRoleLister(metav1.NamespaceAll, roleInformer.Lister())
 	sharedIndexInformers = append(sharedIndexInformers, roleInformer.Informer())
 
-	labelObjects := func(gvr schema.GroupVersionResource, informer cache.SharedIndexInformer, sync queueinformer.LegacySyncHandler) error {
+	complete := map[schema.GroupVersionResource][]bool{}
+	completeLock := &sync.Mutex{}
+
+	labelObjects := func(gvr schema.GroupVersionResource, informer cache.SharedIndexInformer, sync func(done func() bool) queueinformer.LegacySyncHandler) error {
 		if canFilter {
 			return nil
 		}
+		var idx int
```

Comment:
This code hurts my brain. Help me understand it? 😄

Reply:
1. We likely call labelObjects multiple times for the same GVR
2. Keep track of syncs for all GVRs for all invocations
3. For each invocation, when the informer for the GVR has synced, set its spot in `complete` to true

HasSynced func:
1. Grab lock
2. Set the complete spot to true
3. Loop through all tracked syncs by GVR
    1. Loop through all invocations for each GVR
4. If all invocations across all GVRs have synced and marked completed, all GVRs we're labeling **in this file** have been synced

?

Reply:
Oh but it's more subtle than that, because `queueinformer.WithSyncer(sync(func() bool {` is actually calling `sync` and the inline function here is the definition for `done`

Reply:
I see what you're doing now, and why. I get it, but it's hard to grok.

Reply:
Also still trying to grok. I'd suggest adding comments that explain what's going on. This is definitely something I would be re-grok-ing if I have to look at it again later.

Reply:
Added comments

Reply:
Comments really helped! Thanks.